[feature] #2627: Config finalisation (#2706)
hyperledger-iroha/iroha: docs/source/references/config.md

# Iroha Configuration reference

> In this document we provide a reference and detailed descriptions of Iroha's configuration options. Also make sure to read thoroughly an important note[^1] for an explanation of underlying types and default values as this should clear up any confusion about them.

Reviewer: if the info in this note is so important, it shouldn't be a note. I'd put this info in the main body of the text, not in a footnote

Author: The point is, it is a note so that it's present both in the text and linked to by the relevant types, and footnote's the only viable way I know that can do it. I can move the right after the above paragraph, though.

Reviewer: can't it be a separate section of the document that we link to?

Reviewer: the phrasing of the sentence should be changed in any case:

```suggestion
In this document we provide a reference and detailed descriptions of Iroha's configuration options. The options have different underlying types, which are denoted in code as types wrapped in a single `Option<..>` or in a double `Option<Option<..>>`. For the detailed explanation, please refer to the note at the end of this document[^1].
```

I think this also works:

```suggestion
In this document we provide a reference and detailed descriptions of Iroha's configuration options. The options have different underlying types, which are denoted in code as types wrapped in a single `Option<..>` or in a double `Option<Option<..>>`[^1].
```

Author: Agreed, sounds quite harsh when I think about it. Thank you.

# Iroha Configuration reference

> In this document we provide a reference and detailed descriptions of Iroha's configuration options. Also make sure to read thoroughly the following important [section](#configuration-types) for an explanation of underlying types and default values as this should clear up any confusion about them.

Reviewer: see https://github.com/hyperledger/iroha/pull/2706#discussion_r977425829

## Configuration types

> 
> A type wrapped in a single `Option<..>` signifies that in the corresponding `json` block there is a fallback value for this type, and that it only serves as a reference. If a default for such a type has a `null` value, it means that there is no meaningful fallback available for this particular value. All these default values can be freely obtained from a provided [sample configuration file](../../../configs/peer/config.json), but it should only serve as a starting point. If left unchanged, the sample configuration file as is would still fail to build due to it having `null` in place of [public](#public_key) and [private](#private_key) keys as well as [endpoint](#torii.api_url) [URLs](#torii.telemetry_url). These should be provided either by modifying the sample config file or as environment variables. No other overloading of configuration values happens besides reading them from a file and capturing the environment variables. For both types of configuration options wrapped in a single `Option<..>` (i.e. both those that have meaningful defaults and those that have `null`), failure to provide them in any of the above two ways results in an error.

Reviewer: ```suggestion

### `Option<..>` 
A type wrapped in a single `Option<..>` signifies that in the corresponding `json` block there is a fallback value for this type, and that it only serves as a reference. If a default for such a type has a `null` value, it means that there is no meaningful fallback available for this particular value.

All the default values can be freely obtained from a provided [sample configuration file](../../../configs/peer/config.json), but it should only serve as a starting point. If left unchanged, the sample configuration file would still fail to build due to it having `null` in place of [public](#public_key) and [private](#private_key) keys as well as [endpoint](#torii.api_url) [URLs](#torii.telemetry_url). These should be provided either by modifying the sample config file or as environment variables. No other overloading of configuration values happens besides reading them from a file and capturing the environment variables.

For both types of configuration options wrapped in a single `Option<..>` (i.e. both those that have meaningful defaults and those that have `null`), failure to provide them in any of the above two ways results in an error.
```

> `Option<Option<..>>` types should be distinguished from types wrapped in a single `Option<..>`. Only the double option ones are allowed to stay `null`, meaning that **not** providing them in an environment variable or a file will **not** result in an error. Thus only these types are truly optional in the mundane sense of the word. An example of this distinction is genesis [public](#genesis.account_public_key) and [private](#genesis.account_private_key) key. While the first one is a single `Option<..>` wrapped type, the latter is wrapped in `Option<Option<..>>`. This means that the genesis *public* key should always be provided by the user, be it via a file config or an environment variable, whereas the *private* key is only needed for the peer that submits the genesis block, and can be omitted for all others. The same logic goes for other double option fields such as logger file path.

Reviewer: ```suggestion
### `Option<Option<..>>`
`Option<Option<..>>` types should be distinguished from types wrapped in a single `Option<..>`. Only the double option ones are allowed to stay `null`, meaning that **not** providing them in an environment variable or a file will **not** result in an error.

Thus, only these types are truly optional in the mundane sense of the word. An example of this distinction is genesis [public](#genesis.account_public_key) and [private](#genesis.account_private_key) key. While the first one is a single `Option<..>` wrapped type, the latter is wrapped in `Option<Option<..>>`. This means that the genesis *public* key should always be provided by the user, be it via a file config or an environment variable, whereas the *private* key is only needed for the peer that submits the genesis block, and can be omitted for all others. The same logic goes for other double option fields such as logger file path.
```

> Finally, special note about sumeragi fields with `null` as default: only the [`trusted_peers`](#sumeragi.trusted_peers) field out of the three can be initialized via a provided file or an environment variable. The other two, namely [`key_pair`](#sumeragi.key_pair) and [`peer_id`](#sumeragi.peer_id), go through a process of finalization where their values are derived from the corresponding ones in the uppermost Iroha config (for `key_pair`) or the Torii config (in case of [`p2p_addr`](#torii.p2p_addr)). This ensures that these linked fields stay in sync, and prevents the programmer error when different values are provided to these field pairs. Providing either `sumeragi.key_pair` or `sumeragi.peer_id` by hand will result in an error, as it should never be done directly.

Reviewer: ```suggestion
### Sumeragi: default `null` values

A special note about Sumeragi fields with `null` as default: only the [`trusted_peers`](#sumeragi.trusted_peers) field out of the three can be initialized via a provided file or an environment variable.

The other two fields, [`key_pair`](#sumeragi.key_pair) and [`peer_id`](#sumeragi.peer_id), go through a process of finalization where their values are derived from the corresponding ones in the uppermost Iroha config (for `key_pair`) or the Torii config (in case of [`p2p_addr`](#torii.p2p_addr)). This ensures that these linked fields stay in sync, and prevents the programmer error when different values are provided to these field pairs. Providing either `sumeragi.key_pair` or `sumeragi.peer_id` by hand will result in an error, as it should never be done directly.
```

```json
{
  "PUBLIC_KEY": null,

>   "PRIVATE_KEY": null,

Reviewer 2: In the beginning of the section we say that this is default iroha configuration, but this config is non working, because it has missing fields. Is this ok?

Author: There's still quite some work that needs to be done on kagami docs, and there are quite a few places I don't like about the new `config.md`, such as mentioning the types of `ConfigurationProxy`. At the very least, it needs a small explanation at the start that any fields shown as `null` should be user provided. At best, come up with a new workaround for `kagami` instead of deriving `Documented` on `Proxy`. But as for latter, I'm a bit at my wit's end for now.

Author: Just adding a note at the top of kagami docgen in the last push. Surely a makeshift measure.

`BlockSynchronizer` configuration

> Has type `Option<block_sync::ConfigurationProxy>`. Can be configured via environment variable `IROHA_BLOCK_SYNC`

Reviewer 3: We should add an explanatory note to the header, saying that the `Option<bool>` config variables will fall back to a default value if not provided, while the `bool` value must be specified explicitly. Ideally all `Option<…>` segments should point to that (foot)note.

Author: Done a very extensive one.

`BlockSynchronizer` configuration

> Has type `Option<block_sync::ConfigurationProxy>`. Can be configured via environment variable `IROHA_BLOCK_SYNC`. More explanation in [this section](#configuration-types).

Reviewer: ```suggestion
Has type `Option<block_sync::ConfigurationProxy>`. Can be configured via environment variable `IROHA_BLOCK_SYNC`. Refer to [configuration types](#configuration-types) for details.
```

same in other places in this file

Author: As all this is generated from the above one, choosing this wording

}
```

#### `wsv.wasm_runtime_config.fuel_limit`

> The fuel limit determines the maximum number of instructions that can be executed during the execution of a smart contract.

Reviewer: maybe 

```suggestion
The fuel limit determines the maximum number of instructions that can be executed within a smart contract.
```

Has type `Option<u64>`[^1]. Can be configured via environment variable `WASM_FUEL_LIMIT`

```json
1000000
```

#### `wsv.wasm_runtime_config.max_memory`

> Maximum amount of linear memory a given smartcontract can allocate

Reviewer: ```suggestion
Maximum amount of linear memory a given smart contract can allocate.
```

Has type `Option<u32>`[^1]. Can be configured via environment variable `WASM_MAX_MEMORY`

```json
524288000
```

> [^1]: Types wrapped in a single `Option<..>` signify that there is a fallback value for them specified in the corresponding `json` block, but it only serves as a reference. A default `null` for such types means that there is no meaningful fallback for this particular value available. The default peer `config.json` file can serve as a starting point, but there are only two final sources of truth for configuration: all the variables should either be provided via environment or in a file, and no other overloading happens. Failure to provide them in any of the above two ways results in an error.

Reviewer: not sure I understood you correctly, so correct me if I am wrong, but I think this kind of explanation will be more clear to the users:

```suggestion
[^1]: A type wrapped in a single `Option<..>` signifies that in the corresponding `json` block there is a fallback value for this type, and that it only serves as a reference. The default `null` for such types means that there is no meaningful fallback for this particular value available. Note that all variables should either be provided via environment or in a file. Failure to provide them in any of the above two ways results in an error. This means that for configuration options that have types wrapped in a single `Option<..>` with `null` as the default, you *have to* to use environment variables.
```

> 	`Option<Option<..>>` types should be distinguished from types wrapped in singular `Option<..>`. Only the double option ones are allowed to stay `null`, meaning that *not* providing them in an environment variable or a file will not result in an error. Thus only these types are optional in the mundane sense of the word.

Reviewer: ```suggestion
	`Option<Option<..>>` types should be distinguished from types wrapped in a single `Option<..>`. Only the double option ones are allowed to stay `null`, meaning that *not* providing them in an environment variable or a file will not result in an error. Thus only these types are optional in the mundane sense of the word.
```

}
```

#### `wsv.wasm_runtime_config.fuel_limit`

> The fuel limit determines the maximum number of instructions that can be executed during within a smart contract.

Reviewer: ```suggestion
The fuel limit determines the maximum number of instructions that can be executed within a smart contract.
```
